File: 03. Bootloader/7) BLE-UART Bootloader/7) BLE-UART Bootloader/protocol.c
/*
 * protocol.c
 *
 * Created: 2026-06-08 오전 11:08:40
 *  Author: USER
 */ 


#include "protocol.h"
#include "uart.h"



/* ── CRC16 계산 ── */
/*
 * Polynomial: 0x1021
 * Init:       0xFFFF
 * 계산 범위:  MSG_TYPE + DATA_LENGTH + DATA
 */
uint16_t crc16(uint8_t *data, uint16_t len)
{
	uint16_t crc = 0xFFFF;
	for (uint16_t i = 0; i < len; i++) {
		crc ^= ((uint16_t)data[i] << 8);
		for (uint8_t j = 0; j < 8; j++) {
			crc = (crc & 0x8000) ? (crc << 1) ^ 0x1021 : (crc << 1);
		}
	}
	return crc;
}


/*
 * msg_type으로 구분:
 *   MSG_DATA(0x01) → data_length = 256, data 있음
 *   MSG_END (0x02) → data_length = 0,   data 없음
 */
/* ── 262바이트 수신 후 파싱 ── */
uint8_t recv_packet(bl_data_packet_t *pkt)
{
    uint8_t raw[PKT_SIZE];
	
	while(!USART1_Available());
	do {
		raw[0] = USART1_Receive();
	}while(raw[0] != START_CODE);
	
    /* Step 1. 262바이트 전부 수신 */
    for(uint16_t i = 1; i < PKT_SIZE; i++){
        raw[i] = USART1_Receive();
	}
	

    /* Step 1-1. 262바이트 전부 수신 로그 출력 */	
    //for(uint16_t i = 0; i < PKT_SIZE; i++){
		//USART0_Transmit(hex[raw[i] >> 4]);
		//USART0_Transmit(hex[raw[i] & 0xF]);
		//USART0_Transmit(' ');
    //}
	//USART0_Transmit('|');           // 구분자
	//USART0_Transmit(hex[raw[261] >> 4]);
	//USART0_Transmit(hex[raw[261] & 0xF]);
	//USART0_Transmit('|');	
	//
	//USART0_Transmit('\r');
	//USART0_Transmit('\n');	
	
	
    /* Step 2. START_CODE 확인 */
    //if(raw[0] != START_CODE)
        //return NAK;

    /* Step 3. 파싱 */
	// data_length = little endian
    pkt->start_code  = raw[0];
    pkt->msg_type    = raw[1];
    pkt->data_length = (uint16_t)raw[2]
                     | ((uint16_t)raw[3] << 8);
    for(uint16_t i = 0; i < PAGE_SIZE; i++){
        pkt->data[i] = raw[4 + i];
	}

    pkt->crc16 = (uint16_t)raw[260]
               | ((uint16_t)raw[261] << 8);

    /* Step 4. CRC 검증
     * 범위: MSG_TYPE(1) + DATA_LENGTH(2) + DATA(256)
     */
    uint8_t crc_buf[3 + PAGE_SIZE];
    crc_buf[0] = pkt->msg_type;
    crc_buf[1] = raw[2];
    crc_buf[2] = raw[3];
    for(uint16_t i = 0; i < PAGE_SIZE; i++)
        crc_buf[3 + i] = pkt->data[i];

    uint16_t calc = crc16(crc_buf, 3 + PAGE_SIZE);

    if(calc != pkt->crc16)
        return NAK;

    return pkt->msg_type;
    }

File: 04. web app/protocol.js
export const START_CODE = 0xEF;
export const MSG_DATA = 0x01;
export const MSG_END = 0x02;
export const ACK = 0x06;
export const NAK = 0x15;
export const PAGE_SIZE = 256;
export const PKT_SIZE = 262;
export const CHUNK_SIZE = 20;
export const CHUNK_DELAY = 20;
export const MAX_RETRY = 3;
export const ACK_TIMEOUT = 5000;
export const MAX_FIRMWARE_SIZE = 120 * 1024;

export function crc16(data) {
  let crc = 0xFFFF;
  for (const byte of data) {
    crc ^= byte << 8;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc & 0x8000) ? ((crc << 1) ^ 0x1021) : (crc << 1);
      crc &= 0xFFFF;
    }
  }
  return crc;
}

export function makePacket(msgType, pageData = null) {
  if (msgType !== MSG_DATA && msgType !== MSG_END) throw new RangeError('지원하지 않는 메시지 유형입니다.');
  if (pageData && pageData.length > PAGE_SIZE) throw new RangeError('페이지 데이터는 256바이트 이하여야 합니다.');

  const packet = new Uint8Array(PKT_SIZE);
  packet[0] = START_CODE;
  packet[1] = msgType;
  // The bootloader expects the padded 256-byte field for DATA and END packets.
  const dataLength = PAGE_SIZE;
  packet[2] = dataLength & 0xFF;
  packet[3] = dataLength >> 8;
  packet.fill(0xFF, 4, 260);
  if (pageData?.length) packet.set(pageData, 4);

  const crc = crc16(packet.subarray(1, 260));
  packet[260] = crc & 0xFF;
  packet[261] = crc >> 8;
  return packet;
}

export const sleep = (milliseconds) => new Promise(resolve => setTimeout(resolve, milliseconds));

export async function sendChunked(write, packet, shouldCancel = () => false) {
  for (let offset = 0; offset < packet.length; offset += CHUNK_SIZE) {
    if (shouldCancel()) throw new DOMException('전송이 중단되었습니다.', 'AbortError');
    await write(packet.slice(offset, offset + CHUNK_SIZE));
    await sleep(CHUNK_DELAY);
  }
}

export function packetPreview(packet) {
  const hex = bytes => Array.from(bytes, byte => byte.toString(16).padStart(2, '0').toUpperCase()).join(' ');
  return `${hex(packet.slice(0, 10))} ... ${hex(packet.slice(258))}`;
}
